Cython/Utility/StringTools.c

```diff
@@ -1219,3 +1219,25 @@ static CYTHON_INLINE PyObject* __Pyx_PyUnicode_Unicode(PyObject *obj) {
 #define __Pyx_PyObject_Unicode(obj) \
     (likely(PyUnicode_CheckExact(obj)) ? __Pyx_NewRef(obj) : PyObject_Unicode(obj))
 #endif
+
+
+//////////////////// PyStr_Str.proto ////////////////////
+
+static CYTHON_INLINE PyObject* __Pyx_PyStr_Str(PyObject *obj);/*proto*/
+
+//////////////////// PyStr_Str ////////////////////
+
+static CYTHON_INLINE PyObject* __Pyx_PyStr_Str(PyObject *obj) {
+    return __Pyx_NewRef(obj);
```

Comment:
The unicode implementation has a check for Py_None which if true gets wrapped in `PYUNICODE("None")`. I didn't see an equivalent STR macro in (what I assume should be) the CPython source, but I don't required in any case for tests to pass

Reply:
I think this is necessary - when Cython believes something is a fixed Python type it will (usually) also accept `None`.

For `PyObject_Str` that's no problem - if it's passed `None` then the conversion to string happens through the regular mechanism. However it is a problem for this shortcut which only increases the reference count.

Reply:
Wouldn't the branch at line 2329 in Optimize.py ever prevent `None` from coming through here? IIUC this path (at least within the scope of this PR) would only be reached when a "string" object is supplied as the arg

Reply:
No - that's the point I was making. When you type something as `str` it really means `str or None`.

See https://cython.readthedocs.io/en/latest/src/userguide/extension_types.html#extension-types-and-none - I know this is discussed in terms of extension types, but it is the same.

Reply:
Any tips on finding where the PYUNICODE macro is defined?

Reply:
It's not a real macro - it's a preprocessing that Cython does to the utility code https://github.com/cython/cython/blob/e9d7fd45a2e829ee93980b1c0612f8476a46431d/Cython/Compiler/Code.py#L530 .

It doesn't look like there's an equivalent for string, so you might have to make one... maybe... I'm not 100% confident

Reply:
Use `PYIDENT("None")`. It creates a string that Python would accept as an identifier, and that means a `str`, in both Py2 and Py3.